base/libdl.jl

```diff
@@ -314,4 +300,101 @@ function dllist()
     return dynamic_libraries
 end
 
+
+"""
+    LazyLibrary(name, flags = <default dlopen flags>,
+                dependencies = LazyLibrary[], on_load_callback = nothing)
+
+Represents a lazily-loaded library that opens itself and its dependencies on first usage
+in a `dlopen()`, `dlsym()`, or `ccall()` usage.  While this structure contains the
+ability to run arbitrary code on first load via `on_load_callback`, we caution that this
+should be used sparingly, as it is not expected that `ccall()` should result in large
+amounts of Julia code being run.
+"""
+mutable struct LazyLibrary
+    # Name and flags to open with
+    const name::String
+    const flags::UInt32
+
+    # Dependencies that must be loaded before we can load
+    const dependencies::Vector{LazyLibrary}
+
+    # Function that gets called once upon initial load with the pointer as an argument
+    const on_load_callback::Union{Nothing,Function}
+    # Function that gets called once upon final unload with the pointer as an argument
+    const on_unload_callback::Union{Nothing,Function}
+
+    # Pointer that we eventually fill out upon first `dlopen()`
+    @atomic handle::Ptr{Cvoid}
+    @atomic refs::Int
+    function LazyLibrary(name, flags = default_rtld_flags, dependencies = LazyLibrary[],
+                         on_load_callback = nothing, on_unload_callback = nothing)
+        return new(
+            String(name),
+            UInt32(flags),
+            collect(dependencies),
+            on_load_callback,
+            on_unload_callback,
+            C_NULL,
+            0,
+        )
+    end
+end
+
+function dlopen(ll::LazyLibrary, flags::Integer = ll.flags; kwargs...)
+    # Only load once
+    handle = @atomic ll.handle
+    if handle != C_NULL
+        @atomic ll.refs += 1
+        return handle
+    end
+
+    # Ensure that all dependencies are loaded
+    for dep in ll.dependencies
+        dlopen(dep; kwargs...)
+    end
+
+    # Load our library
+    handle = dlopen(ll.name, flags; kwargs...)
+    @atomic ll.handle = handle
+    @atomic ll.refs += 1
+
+    # Invoke our on load callback, if it exists
+    if ll.on_load_callback !== nothing
+        ll.on_load_callback(handle)
+    end
+    return handle
+end
+dlsym(ll::LazyLibrary, args...; kwargs...) = dlsym(dlopen(ll), args...; kwargs...)
+function dlclose(ll::LazyLibrary)
+    if @atomic(ll.refs) > 0 && @atomic(ll.handle) != C_NULL
+        @atomic ll.refs -= 1
+        if @atomic(ll.refs) <= 0
```

Comment:
```suggestion
        refs = @atomic ll.refs -= 1
        if refs <= 0
```

refs should never be `<0`?